### PR TITLE
[posix] adjust posix platform header including

### PR DESCRIPTION
**Pull request by Irving-cl**
This purpose of this PR is to avoid the include chain starting from
`posix/platform/xxx_interface.hpp` from containing anything from OT
core and `lib/platform/exit_code.h`. Because `xxx_interface.hpp` is 
included by `spinel_manager.hpp` which is intended to be used in
`ot-br-posix`.

The PR does these things:
* Move those configs that derived from OT configs into a separate
config file `openthread-posix-core-related-config.h` and only include 
this file when needed. And remove the including of `openthread-core-config.h`
from `openthread-posix-config.h`.
* Remove including of `platform-posix.h` in headers and add necessary
including. i.e., `lib/url/url.hpp`. Because `platform-posix.h` includes
`lib/platform/exit_code.h` which contains some widely used name like
`VerifyOrDie` and may cause conflicts.
* In some posix source files, add including of `openthread-core-config.h`
if necessary. Originally, these configs were included indirectly by
the chain `openthread-posix-config.h -> openthread-core-config.h`

### CODE REVIEW

**Comment by zhanglongxia** [comment]
Is it possible to add new config file `openthread-posix-platform-config.h`, which contains the `OPENTHREAD_POSIX_CONFIG_*` configurations of the file `openthread-posix-config.h`. Then the file `openthread-posix-config.h` contains `openthread-core-config.h` and `openthread-posix-platform-config.h`. The `xxx_interface.hpp` only include the file `openthread-posix-platform-config.h`. This will be more flexible when including header files.

**Comment by zhanglongxia** [comment]
> So maybe let's the devide the options into two categories: related with core and unrelated with core. And let's naming them as `openthread-posix-core-related-config.h` and `openthread-posix-config.h`. Thoughts?

Sounds good.

**Review by zhanglongxia — APPROVED** [review]
LGTM

File: src/posix/platform/infra_if.hpp [hunk 1]
```diff
@@ -35,8 +35,10 @@
 #define OT_POSIX_PLATFORM_INFRA_IF_HPP_
 
 #include "openthread-posix-config.h"
+#include "openthread-posix-core-related-config.h"
```
Comment (abtink) [inline src/posix/platform/infra_if.hpp:38]: @Irving-cl, I understand the goal of making spinel_interface modules independent of the OT code (as a standalone library), but I'm not clear why we need to change infra_if, netif, udp, etc., so they don't include openthread-core-config.h. These modules are part of the platform layer and should be able to see this header. They shouldn't be part of the spinel library module used by a host-side implementation.

File: src/posix/platform/openthread-posix-core-related-config.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ * @def OPENTHREAD_POSIX_CONFIG_INFRA_IF_ENABLE
+ *
+ * Defines `1` to enable the posix implementation of platform/infra_if.h APIs.
+ * The default value is set to `OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE` if it's
+ * not explicit defined.
+ */
+#ifndef OPENTHREAD_POSIX_CONFIG_INFRA_IF_ENABLE
+#define OPENTHREAD_POSIX_CONFIG_INFRA_IF_ENABLE \
+    (OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE || OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE)
+#endif
```
Comment (abtink) [inline src/posix/platform/openthread-posix-core-related-config.h:55]: I believe replicating the configurations, defining similar ones under new name of `OPENTHREAD_POSIX` derived from `OPENTHREAD_CONFIG`, can get confusing and may lead to incorrect usage. Do we need this? Is there any other way?

**Comment by Irving-cl** [comment]
@abtink Thanks! After thinking more on this, I think we should use another alternative.

FYI, the original intention of this PR is to make `posix/platform/spinel_manager.hpp` could be directly included from ot-br-posix (for controller_openthread in the NCP case: https://github.com/openthread/ot-br-posix/pull/2283/files#diff-7b951670ff4aaa492d9bb268de313f2f459c921e3415bee6798ace678f4e3a3bR62). I did it like this instead of calling `otSysMainloopProcess` and `otSysMainloopUpdate` because in that way we need to develop on two repos frequently which will be difficult and spend more time.

However it looks like including posix platform header directly from `ot-br-posix` isn't a good idea. Because the posix platform header should be able to see the config and headers in ot core and that caused an including issue if we include the posix platform header from `ot-br-posix`.

So I think we should still update the `otSysMainloopProcess` and `otSysMainloopUpdate` in ot posix and call them from `ot-br-posix` in the NCP case. Thoughts?

**Comment by Irving-cl** [comment]
Close this PR. Will use another approach.

**Pull request closed without merging**